### ISSUE TITLE
etcd2 proxy doesn't reconnect to other members after one member reboots

### ISSUE THREAD

**Issue by wkruse** [coreos/etcd#6430]
We are running an `etcd2` _v2.3.2_ cluster with 3 members and 9 proxies, CoreOS stable on bare metal. One of the members did a reboot due to CoreOS update from _1068.10.0_ to _1122.2.0_. `fleet` on one of the proxies started rescheduling units because if couldn't read from `etcd2`. It looks like the proxy was trying to connect to the member, that went down. I would expect it to try other members. 

I did find the following in #4606:

>> is there a way to force etcd in proxy=on mode to retry connections against other peers?

> This is a requests from multiple users. We will probably add. But why do you think this is related to the issue you reported

Is it expected behavior in proxy mode?

### `etcd2` node01 (member)

```
Sep 13 02:00:03 node01 etcd2[1982]: failed to dial 8f5b130176fbc095 on stream MsgApp v2 (dial tcp 10.10.226.13:2380: getsockopt: connection refused)
Sep 13 02:00:03 node01 etcd2[1982]: the connection with 8f5b130176fbc095 became inactive
Sep 13 02:00:04 node01 etcd2[1982]: aa1a486e353dd566 [term: 626] received a MsgVote message with higher term from 7bfeb45fee42b84e [term: 627]
Sep 13 02:00:04 node01 etcd2[1982]: aa1a486e353dd566 became follower at term 627
Sep 13 02:00:04 node01 etcd2[1982]: aa1a486e353dd566 [logterm: 626, index: 375981416, vote: 0] voted for 7bfeb45fee42b84e [logterm: 626, index: 375981416] at term 627
Sep 13 02:00:04 node01 etcd2[1982]: raft.node: aa1a486e353dd566 lost leader 8f5b130176fbc095 at term 627
Sep 13 02:00:04 node01 etcd2[1982]: raft.node: aa1a486e353dd566 elected leader 7bfeb45fee42b84e at term 627
Sep 13 02:00:12 node01 etcd2[1982]: the connection to peer 8f5b130176fbc095 is unhealthy
Sep 13 02:00:42 node01 etcd2[1982]: the connection to peer 8f5b130176fbc095 is unhealthy
Sep 13 02:01:12 node01 etcd2[1982]: the connection to peer 8f5b130176fbc095 is unhealthy
Sep 13 02:01:42 node01 etcd2[1982]: the connection to peer 8f5b130176fbc095 is unhealthy
Sep 13 02:02:12 node01 etcd2[1982]: the connection to peer 8f5b130176fbc095 is unhealthy
Sep 13 02:02:42 node01 etcd2[1982]: the connection to peer 8f5b130176fbc095 is unhealthy
Sep 13 02:03:12 node01 etcd2[1982]: the connection to peer 8f5b130176fbc095 is unhealthy
Sep 13 02:03:42 node01 etcd2[1982]: the connection to peer 8f5b130176fbc095 is unhealthy
Sep 13 02:03:48 node01 etcd2[1982]: the connection with 8f5b130176fbc095 became active
```

### `etcd2` node02 (member, became the leader)

```
Sep 13 02:00:03 node02 etcd2[2018]: failed to dial 8f5b130176fbc095 on stream MsgApp v2 (dial tcp 10.10.226.13:2380: getsockopt: connection refused)
Sep 13 02:00:03 node02 etcd2[2018]: the connection with 8f5b130176fbc095 became inactive
Sep 13 02:00:04 node02 etcd2[2018]: 7bfeb45fee42b84e is starting a new election at term 626
Sep 13 02:00:04 node02 etcd2[2018]: 7bfeb45fee42b84e became candidate at term 627
Sep 13 02:00:04 node02 etcd2[2018]: 7bfeb45fee42b84e received vote from 7bfeb45fee42b84e at term 627
Sep 13 02:00:04 node02 etcd2[2018]: 7bfeb45fee42b84e [logterm: 626, index: 375981416] sent vote request to 8f5b130176fbc095 at term 627
Sep 13 02:00:04 node02 etcd2[2018]: 7bfeb45fee42b84e [logterm: 626, index: 375981416] sent vote request to aa1a486e353dd566 at term 627
Sep 13 02:00:04 node02 etcd2[2018]: raft.node: 7bfeb45fee42b84e lost leader 8f5b130176fbc095 at term 627
Sep 13 02:00:04 node02 etcd2[2018]: 7bfeb45fee42b84e received vote from aa1a486e353dd566 at term 627
Sep 13 02:00:04 node02 etcd2[2018]: 7bfeb45fee42b84e [quorum:2] has received 2 votes and 0 vote rejections
Sep 13 02:00:04 node02 etcd2[2018]: 7bfeb45fee42b84e became leader at term 627
Sep 13 02:00:04 node02 etcd2[2018]: raft.node: 7bfeb45fee42b84e elected leader 7bfeb45fee42b84e at term 627
Sep 13 02:00:04 node02 etcd2[2018]: failed to reach the peerURL(http://10.10.226.13:2380) of member 8f5b130176fbc095 (Get http://10.10.226.13:2380/version: dial tcp 10.10.226.13:2380: getsockopt: connection refused)
Sep 13 02:00:04 node02 etcd2[2018]: cannot get the version of member 8f5b130176fbc095 (Get http://10.10.226.13:2380/version: dial tcp 10.10.226.13:2380: getsockopt: connection refused)
...
Sep 13 02:00:10 node02 etcd2[2018]: etcdserver: request timed out, possibly due to previous leader failure
Sep 13 02:00:12 node02 etcd2[2018]: etcdserver: request timed out, possibly due to previous leader failure [merged 35 repeated lines in 1.69s]
...
Sep 13 02:00:17 node02 etcd2[2018]: failed to reach the peerURL(http://10.10.226.13:2380) of member 8f5b130176fbc095 (Get http://10.10.226.13:2380/version: dial tcp 10.10.226.13:2380: i/o timeout)
Sep 13 02:00:17 node02 etcd2[2018]: cannot get the version of member 8f5b130176fbc095 (Get http://10.10.226.13:2380/version: dial tcp 10.10.226.13:2380: i/o timeout)
...
Sep 13 02:00:25 node02 etcd2[2018]: the connection to peer 8f5b130176fbc095 is unhealthy
...
Sep 13 02:03:48 node02 etcd2[2018]: the connection with 8f5b130176fbc095 became active
```

### `etcd2` node03 (member, was leader, rebooted due to CoreOS update)

```
Sep 13 02:00:03 node03 systemd[1]: Stopping etcd2...
Sep 13 02:00:03 node03 etcd2[2048]: received terminated signal, shutting down...
Sep 13 02:00:03 node03 etcd2[2048]: failed to read 7bfeb45fee42b84e on stream MsgApp v2 (net/http: request canceled)
Sep 13 02:00:03 node03 etcd2[2048]: the connection with 7bfeb45fee42b84e became inactive
Sep 13 02:00:03 node03 systemd[1]: Stopped etcd2.
-- Reboot --
Sep 13 02:03:48 node03 systemd[1]: Starting etcd2...
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_ADVERTISE_CLIENT_URLS=http://10.10.226.13:2379
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_DATA_DIR=/var/lib/etcd2
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_INITIAL_ADVERTISE_PEER_URLS=http://10.10.226.13:2380
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_INITIAL_CLUSTER=node01=http://10.10.226.11:2380,node02=http://10.10.226.12:2380,node03=http://10.10.226.13:2380
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_INITIAL_CLUSTER_STATE=new
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_INITIAL_CLUSTER_TOKEN=cf014ad0-cbc7-4138-91db-5eaa54da2ef5
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_LISTEN_PEER_URLS=http://10.10.226.13:2380
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_NAME=node03
Sep 13 02:03:48 node03 etcd2[2434]: recognized and used environment variable ETCD_PROXY=off
Sep 13 02:03:48 node03 etcd2[2434]: etcd Version: 2.3.2
Sep 13 02:03:48 node03 etcd2[2434]: Git SHA: ce63f10
Sep 13 02:03:48 node03 etcd2[2434]: Go Version: go1.5.4
Sep 13 02:03:48 node03 etcd2[2434]: Go OS/Arch: linux/amd64
Sep 13 02:03:48 node03 etcd2[2434]: setting maximum number of CPUs to 32, total number of available CPUs is 32
Sep 13 02:03:48 node03 etcd2[2434]: the server is already initialized as member before, starting as etcd member...
Sep 13 02:03:48 node03 etcd2[2434]: listening for peers on http://10.10.226.13:2380
Sep 13 02:03:48 node03 etcd2[2434]: listening for client requests on http://0.0.0.0:2379
Sep 13 02:03:48 node03 etcd2[2434]: recovered store from snapshot at index 375977587
Sep 13 02:03:48 node03 etcd2[2434]: name = node03
Sep 13 02:03:48 node03 etcd2[2434]: data dir = /var/lib/etcd2
Sep 13 02:03:48 node03 etcd2[2434]: member dir = /var/lib/etcd2/member
Sep 13 02:03:48 node03 etcd2[2434]: heartbeat = 100ms
Sep 13 02:03:48 node03 etcd2[2434]: election = 1000ms
Sep 13 02:03:48 node03 etcd2[2434]: snapshot count = 10000
Sep 13 02:03:48 node03 etcd2[2434]: advertise client URLs = http://10.10.226.13:2379
Sep 13 02:03:48 node03 etcd2[2434]: restarting member 8f5b130176fbc095 in cluster 294917e5a1e04059 at commit index 375981416
Sep 13 02:03:48 node03 etcd2[2434]: 8f5b130176fbc095 became follower at term 626
Sep 13 02:03:48 node03 etcd2[2434]: newRaft 8f5b130176fbc095 [peers: [7bfeb45fee42b84e,8f5b130176fbc095,aa1a486e353dd566], term: 626, commit: 375981416, applied: 375977587, lastindex: 375981416, lastterm: 626]
Sep 13 02:03:48 node03 etcd2[2434]: added member 7bfeb45fee42b84e [http://10.10.226.12:2380] to cluster 294917e5a1e04059 from store
Sep 13 02:03:48 node03 etcd2[2434]: added member 8f5b130176fbc095 [http://10.10.226.13:2380] to cluster 294917e5a1e04059 from store
Sep 13 02:03:48 node03 etcd2[2434]: added member aa1a486e353dd566 [http://10.10.226.11:2380] to cluster 294917e5a1e04059 from store
Sep 13 02:03:48 node03 etcd2[2434]: set the cluster version to 2.3 from store
Sep 13 02:03:48 node03 etcd2[2434]: the connection with 7bfeb45fee42b84e became active
Sep 13 02:03:48 node03 etcd2[2434]: starting server... [version: 2.3.2, cluster version: 2.3]
Sep 13 02:03:48 node03 systemd[1]: Started etcd2.
Sep 13 02:03:48 node03 etcd2[2434]: the connection with aa1a486e353dd566 became active
Sep 13 02:03:48 node03 etcd2[2434]: 8f5b130176fbc095 [term: 626] received a MsgHeartbeat message with higher term from 7bfeb45fee42b84e [term: 627]
Sep 13 02:03:48 node03 etcd2[2434]: 8f5b130176fbc095 became follower at term 627
Sep 13 02:03:48 node03 etcd2[2434]: raft.node: 8f5b130176fbc095 elected leader 7bfeb45fee42b84e at term 627
Sep 13 02:03:48 node03 etcd2[2434]: 8f5b130176fbc095 [commit: 375981416, lastindex: 375981416, lastterm: 626] starts to restore snapshot [index: 375989983, term: 627]
Sep 13 02:03:48 node03 etcd2[2434]: log [committed=375981416, applied=375981416, unstable.offset=375981417, len(unstable.Entries)=0] starts to restore snapshot [index: 375989983, term: 627]
Sep 13 02:03:48 node03 etcd2[2434]: 8f5b130176fbc095 restored progress of 7bfeb45fee42b84e [next = 375989984, match = 0, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
Sep 13 02:03:48 node03 etcd2[2434]: 8f5b130176fbc095 restored progress of 8f5b130176fbc095 [next = 375989984, match = 375989983, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
Sep 13 02:03:48 node03 etcd2[2434]: 8f5b130176fbc095 restored progress of aa1a486e353dd566 [next = 375989984, match = 0, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
Sep 13 02:03:48 node03 etcd2[2434]: 8f5b130176fbc095 [commit: 375989983] restored snapshot [index: 375989983, term: 627]
Sep 13 02:03:48 node03 etcd2[2434]: raft applied incoming snapshot at index 375989983
Sep 13 02:03:48 node03 etcd2[2434]: added member 7bfeb45fee42b84e [http://10.10.226.12:2380] to cluster 294917e5a1e04059 from store
Sep 13 02:03:48 node03 etcd2[2434]: added member 8f5b130176fbc095 [http://10.10.226.13:2380] to cluster 294917e5a1e04059 from store
Sep 13 02:03:48 node03 etcd2[2434]: added member aa1a486e353dd566 [http://10.10.226.11:2380] to cluster 294917e5a1e04059 from store
Sep 13 02:03:48 node03 etcd2[2434]: set the cluster version to 2.3 from store
Sep 13 02:03:48 node03 etcd2[2434]: failed to read 7bfeb45fee42b84e on stream MsgApp v2 (net/http: request canceled)
Sep 13 02:03:48 node03 etcd2[2434]: the connection with 7bfeb45fee42b84e became inactive
Sep 13 02:03:48 node03 etcd2[2434]: failed to read aa1a486e353dd566 on stream Message (net/http: request canceled)
Sep 13 02:03:48 node03 etcd2[2434]: the connection with aa1a486e353dd566 became inactive
Sep 13 02:03:48 node03 etcd2[2434]: recovered from incoming snapshot at index 375989983
Sep 13 02:03:48 node03 etcd2[2434]: the connection with 7bfeb45fee42b84e became active
Sep 13 02:03:48 node03 etcd2[2434]: the connection with aa1a486e353dd566 became active
Sep 13 02:03:48 node03 etcd2[2434]: published {Name:node03 ClientURLs:[http://10.10.226.13:2379]} to cluster 294917e5a1e04059
```

### `etcd2` node10 (proxy)

```
Sep 13 02:00:05 node10 fleetd[2026]: ERROR engine.go:217: Engine leadership lost, renewal failed: context deadline exceeded
Sep 13 02:00:05 node10 etcd2[2016]: proxy: client 10.10.226.20:36926 closed request prematurely
Sep 13 02:00:18 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:18 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:00:18 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:18 node10 fleetd[2026]: ERROR reconcile.go:120: Failed fetching Units from Registry: context deadline exceeded
Sep 13 02:00:18 node10 fleetd[2026]: ERROR reconcile.go:73: Unable to determine agent's desired state: context deadline exceeded
Sep 13 02:00:18 node10 fleetd[2026]: ERROR engine.go:149: Unable to determine cluster engine version
Sep 13 02:00:18 node10 etcd2[2016]: proxy: client 10.10.226.20:36994 closed request prematurely
Sep 13 02:00:18 node10 etcd2[2016]: proxy: client 10.10.226.20:36996 closed request prematurely
Sep 13 02:00:23 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:00:24 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:24 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:00:24 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:24 node10 fleetd[2026]: ERROR engine.go:149: Unable to determine cluster engine version
Sep 13 02:00:24 node10 etcd2[2016]: proxy: client 10.10.226.20:37044 closed request prematurely
Sep 13 02:00:29 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:00:30 node10 fleetd[2026]: ERROR engine.go:149: Unable to determine cluster engine version
Sep 13 02:00:30 node10 etcd2[2016]: proxy: client 10.10.226.20:37050 closed request prematurely
Sep 13 02:00:30 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:30 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:00:30 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:35 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:00:36 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:36 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:00:36 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:41 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:00:43 node10 fleetd[2026]: ERROR reconcile.go:120: Failed fetching Units from Registry: context deadline exceeded
Sep 13 02:00:43 node10 fleetd[2026]: ERROR reconcile.go:73: Unable to determine agent's desired state: context deadline exceeded
Sep 13 02:00:43 node10 etcd2[2016]: proxy: client 10.10.226.20:37100 closed request prematurely
Sep 13 02:00:44 node10 fleetd[2026]: ERROR engine.go:149: Unable to determine cluster engine version
Sep 13 02:00:44 node10 etcd2[2016]: proxy: client 10.10.226.20:37104 closed request prematurely
Sep 13 02:00:45 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: getsockopt: no route to host
Sep 13 02:00:45 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:00:45 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: getsockopt: no route to host
Sep 13 02:00:45 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: getsockopt: no route to host
Sep 13 02:00:50 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:01:18 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:18 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:01:18 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:18 node10 fleetd[2026]: ERROR reconcile.go:120: Failed fetching Units from Registry: context deadline exceeded
Sep 13 02:01:18 node10 fleetd[2026]: ERROR reconcile.go:73: Unable to determine agent's desired state: context deadline exceeded
Sep 13 02:01:18 node10 etcd2[2016]: proxy: client 10.10.226.20:37218 closed request prematurely
Sep 13 02:01:18 node10 fleetd[2026]: ERROR engine.go:149: Unable to determine cluster engine version
Sep 13 02:01:18 node10 etcd2[2016]: proxy: client 10.10.226.20:37222 closed request prematurely
Sep 13 02:01:23 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:01:24 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:24 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:01:24 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:24 node10 fleetd[2026]: ERROR engine.go:149: Unable to determine cluster engine version
Sep 13 02:01:24 node10 etcd2[2016]: proxy: client 10.10.226.20:37268 closed request prematurely
Sep 13 02:01:29 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:01:30 node10 fleetd[2026]: ERROR engine.go:149: Unable to determine cluster engine version
Sep 13 02:01:30 node10 etcd2[2016]: proxy: client 10.10.226.20:37274 closed request prematurely
Sep 13 02:01:32 node10 fleetd[2026]: ERROR engine.go:149: Unable to determine cluster engine version
Sep 13 02:01:32 node10 etcd2[2016]: proxy: client 10.10.226.20:37272 closed request prematurely
Sep 13 02:01:32 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: getsockopt: no route to host
Sep 13 02:01:32 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:01:37 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:01:38 node10 fleetd[2026]: WARN job.go:268: No Unit found in Registry for Job(some-service@1.service)
Sep 13 02:01:38 node10 fleetd[2026]: ERROR job.go:108: Failed to parse Unit from etcd: unable to parse Unit in Registry at key /_coreos.com/fleet/job/some-service@1.service/object
Sep 13 02:01:38 node10 etcd2[2016]: proxy: client 10.10.226.20:37280 closed request prematurely
Sep 13 02:01:38 node10 etcd2[2016]: proxy: client 10.10.226.20:37290 closed request prematurely
Sep 13 02:01:38 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:38 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37298 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37296 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37322 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37300 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37302 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37304 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37306 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37308 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37310 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37312 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37314 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37316 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37318 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37324 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37320 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37326 closed request prematurely
Sep 13 02:01:39 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:39 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:39 node10 etcd2[2016]: proxy: client 10.10.226.20:37370 closed request prematurely
Sep 13 02:01:39 node10 fleetd[2026]: INFO manager.go:138: Triggered systemd unit some-service@1.service stop: job=40007
Sep 13 02:01:39 node10 fleetd[2026]: INFO manager.go:259: Removing systemd unit some-service@1.service
Sep 13 02:01:39 node10 fleetd[2026]: INFO manager.go:182: Instructing systemd to reload units
Sep 13 02:01:39 node10 fleetd[2026]: INFO reconcile.go:330: AgentReconciler completed task: type=UnloadUnit job=some-service@1.service reason="unit loaded but not scheduled here"
Sep 13 02:01:39 node10 fleetd[2026]: INFO reconcile.go:330: AgentReconciler completed task: type=ReloadUnitFiles job=N/A reason="always reload unit files"
Sep 13 02:01:42 node10 fleetd[2026]: INFO manager.go:246: Writing systemd unit some-service@1.service (1612b)
Sep 13 02:01:42 node10 fleetd[2026]: INFO manager.go:182: Instructing systemd to reload units
Sep 13 02:01:42 node10 fleetd[2026]: INFO manager.go:127: Triggered systemd unit some-service@1.service start: job=40308
Sep 13 02:01:42 node10 fleetd[2026]: INFO reconcile.go:330: AgentReconciler completed task: type=LoadUnit job=some-service@1.service reason="unit scheduled here but not loaded"
Sep 13 02:01:42 node10 fleetd[2026]: INFO reconcile.go:330: AgentReconciler completed task: type=ReloadUnitFiles job=N/A reason="always reload unit files"
Sep 13 02:01:42 node10 fleetd[2026]: INFO reconcile.go:330: AgentReconciler completed task: type=StartUnit job=some-service@1.service reason="unit currently loaded but desired state is launched"
Sep 13 02:01:43 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:01:44 node10 etcd2[2016]: proxy: client 10.10.226.20:37378 closed request prematurely
Sep 13 02:01:44 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:44 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:01:44 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:01:45 node10 etcd2[2016]: proxy: client 10.10.226.20:37388 closed request prematurely
Sep 13 02:01:45 node10 fleetd[2026]: WARN engine.go:115: Engine completed reconciliation in 2.041523269s
Sep 13 02:01:49 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:03:47 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: getsockopt: connection refused
Sep 13 02:03:47 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
Sep 13 02:03:47 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: getsockopt: connection refused
Sep 13 02:03:52 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
```

**Comment by wkruse** [comment]
Could be related to https://github.com/coreos/fleet/issues/1289.

**Comment by heyitsanthony** [comment]
@wkruse Where's the part where it can't reconnect? The timestamps show that the proxy notices the down member at 2:00:18 which coincides with the etcdserver stop on `node03` at 2:00:03. The proxy stops complaining at 2:03:52, which coincides with the member coming back online at 2:03:48.

Proxy log (10.10.226.13):
```
Sep 13 02:00:23 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
Sep 13 02:00:24 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: i/o timeout
Sep 13 02:00:24 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 unavailable
...
Sep 13 02:03:47 node10 etcd2[2016]: proxy: failed to direct request to http://10.10.226.13:2379: dial tcp 10.10.226.13:2379: getsockopt: connection refused
Sep 13 02:03:52 node10 etcd2[2016]: proxy: marked endpoint http://10.10.226.13:2379 available, to retest connectivity
```

etcd log (10.10.226.13):
```
Sep 13 02:00:03 node03 systemd[1]: Stopping etcd2...
...
Sep 13 02:03:48 node03 etcd2[2434]: the connection with aa1a486e353dd566 became active
Sep 13 02:03:48 node03 etcd2[2434]: published {Name:node03 ClientURLs:[http://10.10.226.13:2379]} to cluster 294917e5a1e04059
```

**Comment by wkruse** [comment]
@heyitsanthony I guess I misinterpreted those `closed request prematurely` messages. It wasn't clear for me, that those come from connecting to the other members. Thanks for clarification! Looks like it is `fleet` and not `etcd2` problem. I'll close it.

======

After reading https://github.com/coreos/etcd/blob/v2.3.7/Documentation/tuning.md#time-parameters and measuring the RTT

```
--- node01 ping statistics ---
20 packets transmitted, 20 received, 0% packet loss, time 19002ms
rtt min/avg/max/mdev = 0.059/0.106/0.140/0.023 ms
```

I've tuned the `etcd2` and `fleet` time parameters in the cloud config to

```
coreos:
  etcd2:
    heartbeat-interval: 200
    election-timeout: 2000
  fleet:
    engine-reconcile-interval: 10
    etcd-request-timeout: 5
    agent-ttl: "120s"
```

and will try and see if it helps.
